Buddy allocation for large buffers in adaptive allocator

chrisvest opened the pull request:
Motivation:

The histogram bump allocating chunks have poor chunk and memory reuse in practice, which leads to higher memory usage than the pooled allocator whenever an application performs enough allocations of buffers that don't fit in our size-classed chunks.

Buddy allocation should in theory reduce memory consumption by allowing memory reuse within a chunk, similar to the size-classed chunks, but for variable power-of-two sized allocations.

We've found that beyond 16k-20k buffer sizes, allocations predominantly comes in power-of-two sizes, hence buddy allocation should be a good fit for this size regime.

Modification:

* Implement a new chunk type that does buddy allocation, based on an array-embedded binary search tree.
* The tree is encoded as a dense byte array, with two bits marking node or child-node usage, and six bits to encode the node size.
* The histogram pointer-bump allocating chunk implementation is removed, which unlocks potential simplifications and optimizations that will benefit both buddy and size-classed chunks.
* The 32k and 64k size classes are kept for the time being, to keep chunk churn under control, but they are planned to be removed in a follow-up PR.

Result:

We generally get improvements to memory usage, because the buddy allocator is able to reuse its chunks before they are fully deallocated.
If the 32k and 64k size-classes are removed, then the improvements continue to hold up, but we see an increase in allocation churn for buddy chunks.
This needs to be investigated and solved before we can remove the 32k and 64k size-classes.
Presumably, it comes down to making better decisions about the size of the buddy chunks, and in picking which chunks to allocate from next once a magazine has exhausted its current chunk.

chrisvest committed:
First draft of implementing buddy allocation in adaptive for large buffers

chrisvest committed:
Remove the histogram-based bump allocating chunk implementation

chrisvest committed:
Fix a bug in buddy allocation computing incorrect offset for the right side of any subtree

chrisvest committed:
Fix a couple of bugs in the buddy allocation

chrisvest committed:
More accurate remaining capacity for buddy allocation

This is temporary. We should do something faster than this.

chrisvest committed:
Add a data consistency test targeting the buddy allocator

chrisvest committed:
Fix a problem in the buddy allocator

Releasing could end up iterating past its sibling pair, so where releasing a node could end up updating a parent path that wasn't a parent of the released node.

chrisvest committed:
Fix a problem in the buddy allocator

When releasing parent nodes we were only considering the siblings one level down, not whether any other child of a given node had been claimed.
This could lead to grand-parents getting marked as free, if we returned from a child that had a free sibling.
Now we return whether the given subtree is free and only consider the sibling if so.

chrisvest committed:
Remove the Graphviz DOT generating functions

They were used for debugging and are no longer needed.

chrisvest commented:
With the fixes to the freeing code, the buddy allocator is now both faster and use less memory (no longer accidentally keeping memory marked as claimed). However, it's not nearly enough to make up for the removal of the 32k and 64k size classes.
Especially in cases where there's high buffer retention - many buffers alive at the same time. In that case, we end up with many chunks and make poor decisions reusing them because we just a queue per magazine group.

FYI @franz1981

chrisvest commented:
Here's a simulation run with the e-commerce data:
<img width="1512" height="1162" alt="image" src="https://github.com/user-attachments/assets/b89eb7f9-e753-4694-b2ed-50a246074b0d" />
If I comment out the `largeBufferMagazineGroup` path, so large buffers are unpooled instead of using the buddy allocator for pooling, then memory usage is cut in half, which gives us a picture of how far away from the limit we are.

chrisvest committed:
Merge branch '4.2' into 4.2-buddy-alloc

chrisvest committed:
Merge branch '4.2' into 4.2-buddy-alloc

chrisvest committed:
Bring back the 32k and 64k size classes

I think it might be better to look at that problem in a separate follow-up PR.
Don't want each PR to get too big.

chrisvest commented:
@franz1981 I think it'd be better to do chunk picking in a separate PR, so I brought back the 32k and 64k size classes. Those bring the memory usage back to previous levels. I marked this PR ready for review.

chrisvest committed:
Merge branch '4.2' into 4.2-buddy-alloc

# Conflicts:
#	buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

franz1981 reviewed (commented):
I will be soon back after holiday and look at this 🙏

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1471,12 +1284,168 @@ private void releaseSegmentOffsetIntoFreeList(int startIndex) {
+        @Override
+        public boolean readInitInto(AdaptiveByteBuf buf, int size, int startingCapacity, int maxCapacity) {
+            if (!freeList.isEmpty()) {
+                freeList.drain(256, this);
```
Please a comment why we used this "magic number"

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1471,12 +1284,168 @@ private void releaseSegmentOffsetIntoFreeList(int startIndex) {
+        @Override
+        public int remainingCapacity() {
+            if (!freeList.isEmpty()) {
+                freeList.drain(256, this);
```
See above

normanmaurer replied:
Also it feels kind of odd that a "getter" would have such a "side-effect".

chrisvest replied:
Perhaps, but the alternative is to iterate and sum the sizes of pending frees. It'd require adding some reduce-on-range method to the `MpscIntQueue`, which is of course possible.

The whole `remainingCapacity` business is a hold-over from bump allocation, so it's something we should move away from entirely, and instead rely on `readInitInto` returning a `boolean`.

Fixing this up is something I want to do in a future PR, perhaps as part of improving chunk picking.

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1471,12 +1284,168 @@ private void releaseSegmentOffsetIntoFreeList(int startIndex) {
+                } else {
+                    throw new IllegalStateException("The intended segment was not found at index " +
+                            index + ", for size " + size + " and offset " + offset);
+                }
```
No need for the else 
```suggestion
                throw new IllegalStateException("The intended segment was not found at index " +
                        index + ", for size " + size + " and offset " + offset);
```

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1471,12 +1284,168 @@ private void releaseSegmentOffsetIntoFreeList(int startIndex) {
+
+        @Override
+        void releaseSegment(int startingIndex, int size) {
+            int packed = startingIndex / MIN_BUDDY_SIZE | Integer.numberOfTrailingZeros(size / MIN_BUDDY_SIZE) << 16;
```
consider adding some `( )`  to make easer to read

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1471,12 +1284,168 @@ private void releaseSegmentOffsetIntoFreeList(int startIndex) {
+        @Override
+        public void accept(int packed) {
+            // Called by allocating thread when draining freeList.
+            int size = MIN_BUDDY_SIZE << (packed >> 16);
```
Consider using a static field for the `16` and use it here and below to make code easier to read and maintain

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -1247,7 +1062,6 @@ protected void deallocate() {
             } else {
                 RefCnt.resetRefCnt(refCnt);
                 delegate.setIndex(0, 0);
-                allocatedBytes = 0;
```
Is this correct ?

chrisvest replied:
Yes, because updates to this field may be delayed, e.g. by the BuddyChunk freeList. So we can't reset it in case there's relative updates pending. Instead we need to let the concrete Chunk implementations manage it.

normanmaurer reviewed (changes requested):
Generally looking good to me... just a a few nits and a question

chrisvest committed:
Address review comments

chrisvest committed:
BuddyChunks can be released directly back to the group queue

Just like SizeClassedChunks, the BuddyChunks can be released directly back to the shared pool in the magazine group, without waiting for the chunk to be fully freed.
This allows it to be reused much sooner, and reduces memory usage.

This is technically enough to remove the 32k and 64k size classes.
However, if we do that, then it would currently cause an increase in chunk churn, where chunks are allocated and released a lot.
The churn needs to come under control, before those two size classes can be removed.

chrisvest commented:
@franz1981 @normanmaurer Added one more commit where I fixed a missing piece that was causing the increased memory usage seen in https://github.com/netty/netty/pull/16053#issuecomment-3677034782

Like the `SizeClassedChunks`, the `BuddyChunks` can of course be reused before all their buffers have been freed.

This means it's now technically possible to remove the 32k and 64k. However, doing so introduces a fair bit of chunk churn, as can be seen in this chart:
<img width="1124" height="871" alt="image" src="https://github.com/user-attachments/assets/92f318a1-a166-474f-9dcf-7c13c2079e4d" />

Getting the churn under control, and making smarter picking decisions, is what I'll work on next in a separate PR.

normanmaurer commented:
@chrisvest don't we also want to cherry-pick to 4.1 ?

chrisvest commented:
Yeah, I'll add it.
